### PR TITLE
Enforcing NXAPI default HTTP behavior

### PR DESCRIPTION
**Pull request by tstoner**
##### SUMMARY

In the upcoming release of NXOS 9.2 nxapi has changed the default transport from HTTP to HTTPS.  Because of this, test cases that expect HTTP to be enabled by default are failing.

This change-set includes enforcing the default HTTP transport behavior unless explicitly requesting HTTPS.  This ensures backwards compatibility.  Extended nxos_nxapi test cases to cover more HTTPS & HTTP configuration permutations.  Test cases now cover https, https & http and http

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

nxos_nxapi

##### ANSIBLE VERSION

```
ansible 2.7.0.dev0 (devel 0b253df055) last updated 2018/06/21 16:51:18 (GMT -400)
  config file = None
  configured module search path = [u'/Users/tmstoner/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/tmstoner/ansible/ansible/lib/ansible
  executable location = /Users/tmstoner/ansible/ansible/bin/ansible
  python version = 2.7.10 (default, Oct  6 2017, 22:29:07) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.31)]
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

**Comment by mikewiebe** [comment]
Versions Tested and Results:

```
    "msg": "IMAGE VERSION 7.3(0)D1(1)"
    "msg": "IMAGE VERSION 9.2(1)"
    "msg": "IMAGE VERSION 8.3(0)SK(1)"
    "msg": "IMAGE VERSION 7.0(3)F3(4)"
    "msg": "IMAGE VERSION 6.0(2)A8(8)"
    "msg": "IMAGE VERSION 7.0(3)I2(5)"
    "msg": "IMAGE VERSION 7.0(3)I4(7)"
    "msg": "IMAGE VERSION 7.0(3)I5(3)"
    "msg": "IMAGE VERSION 7.0(3)I6(1)"
    "msg": "IMAGE VERSION 7.0(3)I7(3)"

All tests pass
```

**Comment by mikewiebe** [comment]
@trishnaguha and @Qalthos This PR is ready for review.  I have taken this over for @tstoner but am unable to change the title to remove the [WIP] tag since I did not open the original PR.

**NOTE:** Currently for the following case:
```
- name: Enable NXAPI access with default configuration
  nxos_nxapi:
    state: present
```

The module will enforce the default behavior (enable `http`, port `80`) regardless of the default platform behavior but we would like to issue some kind of change notice to indicate that the default behavior will change in a future release to instead (enable `https`, port `443`) regardlesss of the default platform behavior.

What would be the best way to do this?

File: test/integration/targets/prepare_nxos_tests/tasks/main.yml [hunk 1]
```diff
@@ -1,12 +1,17 @@
...
   ignore_errors: yes
 
+- name: Enable Feature NXAPI
+  nxos_nxapi:
+    state: present
+  connection: network_cli
+  ignore_errors: yes
+
 - name: Set nxapi to default state
   nxos_nxapi:
```
Comment (Qalthos) [inline test/integration/targets/prepare_nxos_tests/tasks/main.yml]: You shouldn't need both of these tasks.

File: test/integration/targets/prepare_nxos_tests/tasks/main.yml [hunk 2]
```diff
@@ -1,12 +1,17 @@
...
-      - feature nxapi
       - feature privilege
   connection: network_cli
   ignore_errors: yes
 
+- name: Enable Feature NXAPI
+  nxos_nxapi:
+    state: present
+  connection: network_cli
+  ignore_errors: yes
```
Comment (Qalthos) [inline test/integration/targets/prepare_nxos_tests/tasks/main.yml]: Does this actually emit errors on any platform? I don't see how this could while the next task doesn't

**Comment by Qalthos** [comment]
If I understand correctly, you're looking for something like

```
if not any(module.params['http'], module.params['http_port'], module.params['https'], module.params['https_port']):
    warnings.append([text about default behavior changing])
```

File: lib/ansible/modules/network/nxos/nxos_nxapi.py [hunk 3]
```diff
@@ -171,6 +171,18 @@ def needs_update(x):
...
             return ['no feature nxapi']
         commands.append('feature nxapi')
+    elif want['state'] == 'absent':
+        return commands
+
+    # The latest version (9.2) of NXOS software has changed the platform default
+    # nxapi transfer to https. Enforce the default nxapi http behavior unless
+    # explicitly asking for https.
+
+    if want.get('https') is None or want.get('https') is False:
```
Comment (Qalthos) [inline lib/ansible/modules/network/nxos/nxos_nxapi.py]: I'm not really sure why you are going through all this rather than just setting the `default` key in `argument_spec`... is there something I'm not seeing that is why the defaults aren't set in this module?

**Review by Qalthos — CHANGES_REQUESTED** [review]

File: test/integration/targets/prepare_nxos_tests/tasks/main.yml (hunk 2 above)
Comment (mikewiebe) [reply test/integration/targets/prepare_nxos_tests/tasks/main.yml]: Your right, we don't need `ignore_errors` here.

File: test/integration/targets/prepare_nxos_tests/tasks/main.yml (hunk 1 above)
Comment (mikewiebe) [reply test/integration/targets/prepare_nxos_tests/tasks/main.yml]: Agreed.  Removed.

File: lib/ansible/modules/network/nxos/nxos_nxapi.py (hunk 3 above)
Comment (mikewiebe) [reply lib/ansible/modules/network/nxos/nxos_nxapi.py]: Great suggestion.  Refactoring the code now.

File: lib/ansible/modules/network/nxos/nxos_nxapi.py [hunk 4]
```diff
@@ -279,6 +286,10 @@ def main():
                            supports_check_mode=True)
 
     warnings = list()
+    warning_msg = "Module nxos_nxapi defaults to configure 'http port 80'."
+    warning_msg += "\nDefault behavior is changing in release 2.11 to configure 'https port 443'"
+    warning_msg += " when params 'http, http_port, https, https_port' are not set in the playbook."
+    warnings.append(warning_msg)
```
Comment (mikewiebe) [inline lib/ansible/modules/network/nxos/nxos_nxapi.py]: @Qalthos Now that we are setting defaults in the `argument_spec` it does not make sense to only generate the warning when the parameters are not set.  The warning only occurs the first time the module is called in the playbook so I believe this will achieve the desired result.

**Comment by mikewiebe** [comment]
@Qalthos I believe I have addressed all of your comments.

File: lib/ansible/modules/network/nxos/nxos_nxapi.py [hunk 5]
```diff
@@ -162,37 +162,44 @@ def check_args(module, warnings):
...
-            commands.append('no nxapi https')
+        send_commands.append('feature nxapi')
+    elif want['state'] == 'absent':
+        return send_commands
+
+    # The latest version (9.2) of NXOS software has changed the platform default
+    # nxapi transfer to https. Enforce the default nxapi http behavior unless
+    # explicitly asking for https.
+
+    if needs_update('http'):
```
Comment (Qalthos) [inline lib/ansible/modules/network/nxos/nxos_nxapi.py]: This doesn't cover the case where the protocol is still the same, but the port changed (ie, http_port: 80 to http_port: 8080)

File: lib/ansible/modules/network/nxos/nxos_nxapi.py [hunk 6]
```diff
@@ -162,37 +162,44 @@ def check_args(module, warnings):
...
+        if want.get('http') is False:
+            commands['http'] = 'no nxapi http'
+        if needs_update('http_port') and want.get('http') is True:
+            commands['http'] = 'nxapi http port %s' % want.get('http_port')
+        send_commands.append(commands['http'])
+
+    if needs_update('https'):
```
Comment (Qalthos) [inline lib/ansible/modules/network/nxos/nxos_nxapi.py]: Same here

File: lib/ansible/modules/network/nxos/nxos_nxapi.py [hunk 7]
```diff
@@ -162,37 +162,44 @@ def check_args(module, warnings):
...
-        if want['https'] is True or (want['https'] is None and have['https'] is True):
-            port = want['https_port'] or 443
-            commands.append('nxapi https port %s' % port)
-        elif want['https'] is False:
-            commands.append('no nxapi https')
+        send_commands.append('feature nxapi')
+    elif want['state'] == 'absent':
+        return send_commands
+
+    # The latest version (9.2) of NXOS software has changed the platform default
```
Comment (Qalthos) [inline lib/ansible/modules/network/nxos/nxos_nxapi.py]: Is this comment still meaningful here?

File: lib/ansible/modules/network/nxos/nxos_nxapi.py [hunk 8]
```diff
@@ -162,37 +162,44 @@ def check_args(module, warnings):
 
 
 def map_obj_to_commands(want, have, module):
-    commands = list()
+    send_commands = list()
+    commands = dict()
```
Comment (Qalthos) [inline lib/ansible/modules/network/nxos/nxos_nxapi.py:166]: I'm not understanding what the point of changing commands to a dict here is, can you explain why the change?

**Review by Qalthos — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/nxos/nxos_nxapi.py (hunk 8 above)
Comment (mikewiebe) [reply lib/ansible/modules/network/nxos/nxos_nxapi.py:166]: it's not explicitly necessary but I decided to create a dictionary that I could index with a meaningful key to keep track of the commands that we are building for `http, http_port, https, https_port and sandbox` settings.  This way I can just overwrite the dictionary entry as more information is discovered and then build the final list called `send_commands`.

File: lib/ansible/modules/network/nxos/nxos_nxapi.py (hunk 7 above)
Comment (mikewiebe) [reply lib/ansible/modules/network/nxos/nxos_nxapi.py]: With the last round of changes I suppose not.  I will remove it.

File: lib/ansible/modules/network/nxos/nxos_nxapi.py (hunk 5 above)
Comment (mikewiebe) [reply lib/ansible/modules/network/nxos/nxos_nxapi.py]: *sigh*  Good catch.

File: lib/ansible/modules/network/nxos/nxos_nxapi.py (hunk 4 above)
Comment (trishnaguha) [reply lib/ansible/modules/network/nxos/nxos_nxapi.py]: @mikewiebe The way to do this is to use `module.deprecate`
```
if condition:
    module.deprecate(msg="deprecation msg",
                                  version="2.11")
```
This will generate standard deprecation warning like Ansible does.
I think you can actually eliminate the condition part here as default is set in argspec.

Comment (mikewiebe) [reply lib/ansible/modules/network/nxos/nxos_nxapi.py]: Will do

File: lib/ansible/modules/network/nxos/nxos_nxapi.py (hunk 5 above)
Comment (mikewiebe) [reply lib/ansible/modules/network/nxos/nxos_nxapi.py]: My latest set of diffs address this and I also added a test case to change the http and https port.

**Comment by mikewiebe** [comment]
@Qalthos @trishnaguha Addressed latest review comments.

**Review by trishnaguha — APPROVED** [review]

**Comment by mikewiebe** [comment]
@Qalthos Thanks for the thorough review!  If there are no additional comments can we merge this today?  Thanks!

**Review by Qalthos — APPROVED** [review]

**Comment by mikewiebe** [comment]
@trishnaguha @gundalow Can we get these changes cherry-picked into 2.6 and possibly 2.5?
